test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/CompilerCacheTest.cs

```diff
@@ -334,6 +336,113 @@ public void ConstructorNormalizesPrecompiledViewPath(string viewPath)
             Assert.Same(expected, result.CompilationResult.CompiledType);
         }
 
+        [Fact]
+        public async Task GetOrAdd_AllowsConcurrentCompilationOfMultipleRazorPages()
+        {
+            // Arrange
+            var waitDuration = TimeSpan.FromSeconds(20);
+            var fileProvider = new TestFileProvider();
+            fileProvider.AddFile("/Views/Home/Index.cshtml", "Index content");
+            fileProvider.AddFile("/Views/Home/About.cshtml", "About content");
+            var resetEvent1 = new AutoResetEvent(initialState: false);
+            var resetEvent2 = new ManualResetEvent(initialState: false);
+            var cache = new CompilerCache(fileProvider);
+            var compilingOne = false;
+            var compilingTwo = false;
+
+            // Act
+            var task1 = Task.Run(() =>
+            {
+                return cache.GetOrAdd("/Views/Home/Index.cshtml", file =>
+                {
+                    compilingOne = true;
+
+                    // Event 2
+                    resetEvent1.WaitOne(waitDuration);
+
+                    // Event 3
+                    resetEvent2.Set();
+
+                    // Event 6
+                    resetEvent1.WaitOne(waitDuration);
+
+                    Assert.True(compilingTwo);
+                    return new CompilationResult(typeof(TestView));
+                });
+            });
+
+            var task2 = Task.Run(() =>
+            {
+                // Event 4
+                return cache.GetOrAdd("/Views/Home/About.cshtml", file =>
+                {
+                    compilingTwo = true;
+
+                    // Event 4
+                    resetEvent2.WaitOne(waitDuration);
```

Comment:
Rather than doing waits with timeouts here, we should just unblock the events manually.